File: wrappers/vb/LibRangeMap/TextRangeMapper.vb
Imports System.Text
Imports System.Text.Json

Public Class TextRangeMapper
    Private Shared ReadOnly JsonOptions As New JsonSerializerOptions With {
        .PropertyNamingPolicy = Nothing,
        .WriteIndented = False
    }

    Private ReadOnly _mode As String
    Private ReadOnly _alphabet As String
    Private ReadOnly _outputMin As Double
    Private ReadOnly _outputMax As Double
    Private ReadOnly _allowEmpty As Boolean

    Public Sub New(Optional mode As String = "codepoint", Optional alphabet As String = "", Optional outputMin As Double = -1.0, Optional outputMax As Double = 1.0, Optional allowEmpty As Boolean = False)
        If String.IsNullOrWhiteSpace(mode) Then
            Throw New ArgumentException("mode must not be empty.", NameOf(mode))
        End If
        If Double.IsNaN(outputMin) OrElse Double.IsInfinity(outputMin) OrElse Double.IsNaN(outputMax) OrElse Double.IsInfinity(outputMax) Then
            Throw New ArgumentException("output range endpoints must be finite.")
        End If
        If outputMin >= outputMax Then
            Throw New ArgumentException("outputMin must be less than outputMax.")
        End If

        Dim normalizedMode = mode.Trim().ToLowerInvariant()
        If normalizedMode <> "codepoint" AndAlso normalizedMode <> "byte" AndAlso normalizedMode <> "alphabet" Then
            Throw New ArgumentException($"unsupported text mode '{mode}'.", NameOf(mode))
        End If

        If normalizedMode = "alphabet" Then
            If String.IsNullOrEmpty(alphabet) Then
                Throw New ArgumentException("alphabet must not be empty for alphabet mode.", NameOf(alphabet))
            End If
            Dim seen As New HashSet(Of Char)()
            For Each ch In alphabet
                If Not seen.Add(ch) Then
                    Throw New ArgumentException($"duplicate alphabet token '{ch}'.", NameOf(alphabet))
                End If
            Next
        ElseIf alphabet <> "" Then
            Throw New ArgumentException("alphabet must be empty unless mode is alphabet.", NameOf(alphabet))
        End If

        _mode = normalizedMode
        _alphabet = alphabet
        _outputMin = outputMin
        _outputMax = outputMax
        _allowEmpty = allowEmpty
    End Sub

    Public Function MapValue(value As String) As Double()
        If value Is Nothing Then
            Throw New ArgumentNullException(NameOf(value))
        End If
        If value.Length = 0 AndAlso Not _allowEmpty Then
            Throw New ArgumentException("empty text input is invalid by default.", NameOf(value))
        End If

        If _mode = "byte" Then
            Dim bytes = Encoding.UTF8.GetBytes(value)
            If bytes.Length = 0 AndAlso Not _allowEmpty Then
                Throw New ArgumentException("empty text input is invalid by default.", NameOf(value))
            End If
            Dim mapped(bytes.Length - 1) As Double
            For index = 0 To bytes.Length - 1
                mapped(index) = _outputMin + ((CDbl(bytes(index)) / 255.0) * (_outputMax - _outputMin))
            Next
            Return mapped
        End If

        Dim runes = value.EnumerateRunes().ToArray()
        If runes.Length = 0 AndAlso Not _allowEmpty Then
            Throw New ArgumentException("empty text input is invalid by default.", NameOf(value))
        End If

        Dim mappedValues(runes.Length - 1) As Double
        If _mode = "alphabet" Then
            For index = 0 To runes.Length - 1
                Dim ch = ChrW(runes(index).Value)
                Dim position = _alphabet.IndexOf(ch)
                If position < 0 Then
                    Throw New ArgumentException($"unknown text token '{ch}'.", NameOf(value))
                End If
                If _alphabet.Length = 1 Then
                    mappedValues(index) = (_outputMin + _outputMax) / 2.0
                Else
                    mappedValues(index) = _outputMin + ((CDbl(position) / CDbl(_alphabet.Length - 1)) * (_outputMax - _outputMin))
                End If
            Next
            Return mappedValues
        End If

        For index = 0 To runes.Length - 1
            Dim current = CDbl(runes(index).Value)
            mappedValues(index) = _outputMin + ((current / 255.0) * (_outputMax - _outputMin))
        Next
        Return mappedValues
    End Function

    Public Function Map(value As String) As Double()
        Return MapValue(value)
    End Function

    Public Function Spec() As TextMapperSpec
        Return New TextMapperSpec With {
            .SpecVersion = CurrentSpecVersion,
            .MapperType = "text_range",
            .Mode = _mode,
            .Alphabet = _alphabet,
            .OutputRange = New Double() {_outputMin, _outputMax},
            .AllowEmpty = _allowEmpty
        }
    End Function

    Public Function ToJson() As String
        Return JsonSerializer.Serialize(Spec(), JsonOptions)
    End Function

    Public Shared Function FromJson(json As String) As TextRangeMapper
        Dim spec = JsonSerializer.Deserialize(Of TextMapperSpec)(json, JsonOptions)
        If spec Is Nothing Then
            Throw New InvalidOperationException("mapper spec JSON produced null.")
        End If
        Return FromSpec(spec)
    End Function

    Public Shared Function FromSpec(spec As TextMapperSpec) As TextRangeMapper
        If spec.SpecVersion <> CurrentSpecVersion Then
            Throw New ArgumentException($"unsupported spec_version '{spec.SpecVersion}'.", NameOf(spec))
        End If
        If spec.MapperType <> "text_range" Then
            Throw New ArgumentException($"unsupported mapper_type '{spec.MapperType}'.", NameOf(spec))
        End If
        If spec.OutputRange Is Nothing OrElse spec.OutputRange.Length <> 2 Then
            Throw New ArgumentException("output_range must contain exactly two values.", NameOf(spec))
        End If
        Return New TextRangeMapper(spec.Mode, spec.Alphabet, spec.OutputRange(0), spec.OutputRange(1), spec.AllowEmpty)
    End Function
End Class

File: wrappers/csharp/LibRangeMap/TextRangeMapper.cs
using System;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LibRangeMap;

public sealed class TextRangeMapper
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = null,
        WriteIndented = false
    };

    private const string SpecVersion = MapperSpec.CurrentSpecVersion;
    private const string SpecType = "text_range";

    public string Mode { get; }
    public long InputMin { get; }
    public long InputMax { get; }
    public double OutputMin { get; }
    public double OutputMax { get; }
    public bool Clip { get; }
    public bool AllowEmpty { get; }
    public string? Alphabet { get; }
    public string? Name { get; }

    public TextRangeMapper(
        string mode = "codepoint",
        string? alphabet = null,
        double outputMin = -1.0,
        double outputMax = 1.0,
        bool clip = false,
        bool allowEmpty = false,
        string? name = null)
    {
        ValidateFinite(nameof(outputMin), outputMin);
        ValidateFinite(nameof(outputMax), outputMax);

        if (outputMax <= outputMin)
        {
            throw new ArgumentException("output_range must be ordered from lower value to higher value.");
        }
        if (name is not null && name.Length == 0)
        {
            throw new ArgumentException("name must not be empty.", nameof(name));
        }
        if (mode != "codepoint" && mode != "alphabet" && mode != "byte")
        {
            throw new ArgumentException("mode must be one of 'codepoint', 'alphabet', or 'byte'.");
        }

        Mode = mode;
        Clip = clip;
        AllowEmpty = allowEmpty;
        Alphabet = mode == "alphabet" ? NormalizeAlphabet(alphabet) : null;
        OutputMin = outputMin;
        OutputMax = outputMax;
        Name = name;

        if (Mode == "alphabet")
        {
            if (Alphabet is null)
            {
                throw new ArgumentException("alphabet is required for alphabet mode.");
            }

            if (Alphabet.Length < 2)
            {
                throw new ArgumentException("alphabet mode requires at least 2 symbols.");
            }

            InputMin = 0;
            InputMax = Alphabet.Length - 1;
            return;
        }

        if (Mode == "byte")
        {
            InputMin = 0;
            InputMax = 255;
            return;
        }

        InputMin = 0;
        InputMax = 0x10FFFF;
    }

    public double MapValue(char value)
    {
        if (Mode == "alphabet")
        {
            return MapSymbol(new string(value, 1));
        }

        return MapCodePoint((long)value);
    }

    public double[] Map(string value)
    {
        if (value is null)
        {
            throw new ArgumentNullException(nameof(value), "text payload must not be null.");
        }
        if (value.Length == 0 && !AllowEmpty)
        {
            throw new ArgumentException("text payload must not be empty unless allow_empty is true.");
        }

        if (Mode == "byte")
        {
            var bytes = Encoding.UTF8.GetBytes(value);
            var output = new double[bytes.Length];
            for (var i = 0; i < bytes.Length; i++)
            {
                output[i] = MapUnit(bytes[i]);
            }
            return output;
        }

        if (Mode == "alphabet")
        {
            var output = new double[value.Length];
            for (var i = 0; i < value.Length; i++)
            {
                output[i] = MapSymbol(value[i].ToString());
            }
            return output;
        }

        var units = new int[value.Length];
        var idx = 0;
        var runeCount = 0;
        while (idx < value.Length)
        {
            var rune = Rune.GetRuneAt(value, idx);
            units[runeCount++] = rune.Value;
            idx += rune.Utf16SequenceLength;
        }

        var mapped = new double[runeCount];
        for (var i = 0; i < runeCount; i++)
        {
            mapped[i] = MapUnit(units[i]);
        }

        return mapped;
    }

    public string ToJson() => JsonSerializer.Serialize(ToSpec(), JsonOptions);

    public static TextRangeMapper FromJson(string json)
    {
        var spec = JsonSerializer.Deserialize<TextMapperSpec>(json, JsonOptions)
            ?? throw new InvalidOperationException("mapper spec JSON produced null.");
        return FromSpec(spec);
    }

    public static TextRangeMapper FromSpec(TextMapperSpec spec)
    {
        if (spec.SpecVersion != SpecVersion)
        {
            throw new ArgumentException($"unsupported spec_version '{spec.SpecVersion}'.", nameof(spec));
        }
        if (spec.MapperType != SpecType)
        {
            throw new ArgumentException($"unsupported mapper_type '{spec.MapperType}'.", nameof(spec));
        }
        if (spec.OutputRange is null || spec.OutputRange.Length != 2)
        {
            throw new ArgumentException("output_range must contain exactly two values.", nameof(spec));
        }
        if (spec.Mode is null)
        {
            throw new ArgumentException("mode is required.", nameof(spec));
        }
        if (spec.InputRange is null || spec.InputRange.Length != 2)
        {
            throw new ArgumentException("input_range must contain exactly two values.", nameof(spec));
        }
        if (spec.Mode == "alphabet")
        {
            if (spec.Alphabet is null || spec.Alphabet.Length < 2)
            {
                throw new ArgumentException("alphabet mode requires an alphabet with at least 2 symbols.", nameof(spec));
            }
            if (spec.InputRange[0] != 0 || spec.InputRange[1] != spec.Alphabet.Length - 1)
            {
                throw new ArgumentException("input_range does not match alphabet length.", nameof(spec));
            }
        }
        if (spec.Mode == "byte" && (spec.InputRange[0] != 0 || spec.InputRange[1] != 255))
        {
            throw new ArgumentException("input_range must be [0,255] for byte mode.", nameof(spec));
        }
        if (spec.Mode == "codepoint" && (spec.InputRange[0] != 0 || spec.InputRange[1] != 0x10FFFF))
        {
            throw new ArgumentException("input_range must be [0,1114111] for codepoint mode.", nameof(spec));
        }
        if (spec.Mode != "codepoint" && spec.Mode != "alphabet" && spec.Mode != "byte")
        {
            throw new ArgumentException("mode must be one of 'codepoint', 'alphabet', or 'byte'.", nameof(spec));
        }

        return new TextRangeMapper(
            spec.Mode,
            spec.Alphabet,
            spec.OutputRange[0],
            spec.OutputRange[1],
            spec.Clip,
            spec.AllowEmpty,
            spec.Name);
    }

    private double MapSymbol(string symbol)
    {
        if (Alphabet is null)
        {
            throw new InvalidOperationException("alphabet is not configured.");
        }

        var index = Alphabet.IndexOf(symbol, StringComparison.Ordinal);
        if (index < 0)
        {
            throw new KeyNotFoundException($"unknown symbol '{symbol}' for alphabet mode.");
        }

        return MapUnit(index);
    }

    private double MapCodePoint(long value)
    {
        return MapUnit(value);
    }

    private double MapUnit(long value)
    {
        var v = (double)value;

        if (value < InputMin)
        {
            if (!Clip)
            {
                throw new ArgumentOutOfRangeException(nameof(value), $"value {v} is below input_range lower bound {InputMin}; enable clip to clamp.");
            }
            v = InputMin;
        }
        else if (value > InputMax)
        {
            if (!Clip)
            {
                throw new ArgumentOutOfRangeException(nameof(value), $"value {v} is above input_range upper bound {InputMax}; enable clip to clamp.");
            }
            v = InputMax;
        }

        return OutputMin + ((v - InputMin) / (InputMax - InputMin)) * (OutputMax - OutputMin);
    }

    private static string? NormalizeAlphabet(string? alphabet)
    {
        if (alphabet is not null && alphabet.Length == 0)
        {
            throw new ArgumentException("alphabet must not be empty.", nameof(alphabet));
        }

        return alphabet;
    }

    private static void ValidateFinite(string label, double value)
    {
        if (double.IsNaN(value) || double.IsInfinity(value))
        {
            throw new ArgumentException($"{label} must be a finite number.");
        }
    }

    private TextMapperSpec ToSpec()
    {
        return new TextMapperSpec
        {
            SpecVersion = SpecVersion,
            MapperType = SpecType,
            Mode = Mode,
            InputRange = new[] { InputMin, InputMax },
            OutputRange = new[] { OutputMin, OutputMax },
            Clip = Clip,
            AllowEmpty = AllowEmpty,
            Alphabet = Alphabet,
            Name = Name
        };
    }
}

public sealed class TextMapperSpec
{
    [JsonPropertyName("spec_version")]
    public string SpecVersion { get; init; } = MapperSpec.CurrentSpecVersion;

    [JsonPropertyName("mapper_type")]
    public string MapperType { get; init; } = "text_range";

    [JsonPropertyName("mode")]
    public string? Mode { get; init; }

    [JsonPropertyName("input_range")]
    public long[] InputRange { get; init; } = new long[2];

    [JsonPropertyName("output_range")]
    public double[] OutputRange { get; init; } = new double[2];

    [JsonPropertyName("clip")]
    public bool Clip { get; init; }

    [JsonPropertyName("allow_empty")]
    public bool AllowEmpty { get; init; }

    [JsonPropertyName("alphabet")]
    public string? Alphabet { get; init; }

    [JsonPropertyName("name")]
    public string? Name { get; init; }
}
